### ISSUE TITLE
[FR] Warn if composables are used in the wrong places

### ISSUE THREAD

**Issue by wenfangdu** [vuejs/eslint-plugin-vue#2093]
**Please describe what the rule should do:**

> [Composables should only be called synchronously in <script setup> or the setup() hook. In some cases, you can also call them in lifecycle hooks like onMounted().](https://vuejs.org/guide/reusability/composables.html#conventions-and-best-practices)

It would be great to have a rule to check for this like [`react-hooks/rules-of-hooks`](https://reactjs.org/docs/hooks-rules.html#eslint-plugin).

**What category should the rule belong to?**

- [ ] Enforces code style (layout)
- [x] Warns about a potential error (problem)
- [ ] Suggests an alternate way of doing something (suggestion)
- [ ] Other (please specify:)

**Provide 2-3 code examples that this rule should warn about:**

```js
import { useFoo } from '@/composables'

const onClick = async () => useFoo() // 👈 should be warned
```

**Additional context**

NA

**Comment by FloEdelmann** [comment]
How should composables be detected? Any function called `useXyz`?

**Comment by wenfangdu** [comment]
> [It is a convention to name composable functions with camelCase names that start with "use".](https://vuejs.org/guide/reusability/composables.html#naming)

@FloEdelmann And exclude builtin APIs like `useAttrs()`, `useSlots()` etc, since ["They can be used in normal composition API functions as well"](https://vuejs.org/api/sfc-script-setup.html#useslots-useattrs).

**Comment by deleteme** [comment]
React gets away with it in their "rules of hooks" eslint tool. It's a strong convention, so a simple implementation is workable.

https://github.com/facebook/react/blob/b4cdd3e8922713f8c9817b004a0dc51be47bc5df/packages/eslint-plugin-react-hooks/src/RulesOfHooks.js#L13-L23

**Comment by mukundshah** [comment]
One way would be to check if the composable is making use of `getCurrentInstance()` at any level. But then the question arises, how are we going to do that?
